test/hotspot/jtreg/compiler/c2/gvn/ModINodeValueTests.java

```diff
@@ -0,0 +1,199 @@
+
+import compiler.lib.generators.Generator;
+import compiler.lib.generators.Generators;
+import compiler.lib.ir_framework.DontCompile;
+import compiler.lib.ir_framework.IR;
+import compiler.lib.ir_framework.IRNode;
+import compiler.lib.ir_framework.Run;
+import compiler.lib.ir_framework.Test;
+import compiler.lib.ir_framework.TestFramework;
+import jdk.test.lib.Asserts;
+
+/*
+ * @test
+ * @bug 8356813
+ * @summary Test that Value method of ModINode is working as expected.
+ * @library /test/lib /
+ * @run driver compiler.c2.gvn.ModINodeValueTests
+ */
+public class ModINodeValueTests {
+    private static final Generator<Integer> G = Generators.G.ints();
+
+    public static void main(String[] args) {
+        TestFramework.run();
+    }
+
+    @Run(test = {
+        "nonNegativeDividend", "nonNegativeDividendInRange",
+        "negativeDividend", "negativeDividendInRange",
+        "modByKnownBoundsUpper", "modByKnownBoundsUpperInRange",
+        "modByKnownBoundsLower", "modByKnownBoundsLowerInRange",
+        "modByKnownBoundsLimitedByDividendUpper", "modByKnownBoundsLimitedByDividendUpperInRange",
+        "modByKnownBoundsLimitedByDividendLower", "modByKnownBoundsLimitedByDividendLowerInRange"
+    })
+    public void runMethod() {
+        int a = G.next();
+        int b = G.next();
+
+        int min = Integer.MIN_VALUE;
+        int max = Integer.MAX_VALUE;
+
+        assertResult(0, 0);
+        assertResult(a, b);
+        assertResult(min, min);
+        assertResult(max, max);
+    }
+
+    @DontCompile
+    public void assertResult(int x, int y) {
+        Asserts.assertEQ(x != 0 && 123 % x < 0, nonNegativeDividend(x));
+        Asserts.assertEQ(x != 0 && 123 % x <= 0, nonNegativeDividendInRange(x));
+        Asserts.assertEQ(x != 0 && -123 % x > 0, negativeDividend(x));
+        Asserts.assertEQ(x != 0 && -123 % x >= 0, negativeDividendInRange(x));
+        Asserts.assertEQ(x % (((byte) y) + 129) > 255, modByKnownBoundsUpper(x, y));
+        Asserts.assertEQ(x % (((byte) y) + 129) >= 255, modByKnownBoundsUpperInRange(x, y));
+        Asserts.assertEQ(x % (((byte) y) + 129) < -255, modByKnownBoundsLower(x, y));
+        Asserts.assertEQ(x % (((byte) y) + 129) <= -255, modByKnownBoundsLowerInRange(x, y));
+        Asserts.assertEQ(((byte) x) % (((char) y) + 1) > 127, modByKnownBoundsLimitedByDividendUpper(x, y));
+        Asserts.assertEQ(((byte) x) % (((char) y) + 1) >= 127, modByKnownBoundsLimitedByDividendUpperInRange(x, y));
+        Asserts.assertEQ(((byte) x) % (((char) y) + 1) < -128, modByKnownBoundsLimitedByDividendLower(x, y));
+        Asserts.assertEQ(((byte) x) % (((char) y) + 1) <= -128, modByKnownBoundsLimitedByDividendLowerInRange(x, y));
+    }
+
+    @Test
+    @IR(failOn = {IRNode.MOD_I, IRNode.CMP_I})
+    // The sign of the result of % is the same as the sign of the dividend,
+    // i.e., posVal % x < 0 => false.
+    public boolean nonNegativeDividend(int x) {
+        return x != 0 && 123 % x < 0;
```

Comment:
It would be nice to have random dividends in the tests as well to further increase coverage. To still get constant folding, you can put random values in final fields.